sdk/eventhub/Azure.Messaging.EventHubs.Shared/src/Core/EventHubsExceptionExtensions.cs

```diff
@@ -0,0 +1,35 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Azure.Messaging.EventHubs.Core
+{
+    /// <summary>
+    ///   The set of extension methods for the <see cref="EventHubsException" /> class.
+    /// </summary>
+    ///
+    internal static class EventHubsExceptionExtensions
```

Comment:
Echoing a thought that I had elsewhere, maybe instead of extending the failure reason, we could consider another extension method for the associated operation and then could filter based on operation and extract the associated data?

Reply:
I like it, great suggestion!

Reply:
Just to clarify and summarize the various thoughts and suggestions above, here are the options I thing we've raised:

1. Operation specific extension methods
   - No additional `FailureReason`
   - getter/setter extension methods abstract away any notion of key selection and simply get and set an operation specific payload.
   - ex. ` public static string GetPartitionLoadBalancingFailureData(this EventHubsException instance)` and ` public static void SetPartitionLoadBalancingFailureData(this EventHubsException instance, string partitionId)`
2. Data keyed on operation resource
   - No additional `FailureReason`
   - Key the Data based on a known resource such as `Resources.OperationClaimOwnership`
   - ex. ` public static T GetOperationFailureData(this EventHubsException instance, string resource)` and `public static void SetOperationFailureData<T>(this EventHubsException instance, string resource, T data)`
3. Failure code bounded Data values (existing solution)
   - Additional `FailureReason`
   - Key the data based on `FailureReason`

Did I miss or misinterpret any?

Reply:
As usual, thanks for being flexible and talking things through.   Your summary sounds great to me.    I may suggest a very slight twist:

- `public static T GetFailureData<T>(this EventHubsException instance)`
- `public static string GetFailureOperation(this EventHubsException instance)`

_(and, obviously, the associated setter)_

Which would just flip things a bit where a failure could set any kind of data but we assume there is a string that describes an operation.... which is something that is consistent within the core and processor libraries.   Either way, I'm good with the approach.

Reply:
I think I'm missing something obvious here, but in your example methods above how does the get operation know which operation is relevant? Is this meant to be just a single example of some more specifically named methods per operation as in # 1 above?

Reply:
> The general thing that I was looking to do was make "failure data" some object of unknown type that is set for the failure and "operation" a string value that corresponds to the text labels that we set for logging and such to offer context for the failure.

I believe we may be saying the same thing, but I just want to confirm 😄.  I think my confusion is that your description sounds consistent with # 2 above where the "operation" string is accepted in both the setter and the getter. The caller is then responsible to define or use a consistent operation identifier to key the Data. What threw me off is that your method example didn't accept an "operation".

Reply:
I don't think we're saying the same.  :smile:

I wasn't thinking to key off the operation, I was thinking the option would be the value.    So the `Data` dictionary would look something like:

```
Exception.Data
{
    { "FailureData",  PartitionId },
    { "FailureOeration", Resouces.LoadBalancerClaimFailure }
}
```

Reply:
Ah, sorry for the confusion - I get what you are saying now.  So essentially the data dictionary becomes a self describing payload defining the relevant operation and the corresponding FailureData. Assuming I'm not still confused, I'll incorporate this concept into the next commit 😄.

Reply:
Not at all, the problem was all me not being able to explain what was in my head.